packages/repository/src/relations/has-many/has-many-through-repository.factory.ts

```diff
@@ -0,0 +1,219 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/example-todo
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as debugFactory from 'debug';
+import {camelCase} from 'lodash';
+import {DataObject} from '../../common-types';
+import {EntityNotFoundError, InvalidRelationError} from '../../errors';
+import {Entity} from '../../model';
+import {EntityCrudRepository} from '../../repositories/repository';
+import {constrainFilter} from '../../repositories/constraint-utils';
+import {isTypeResolver} from '../../type-resolver';
+import {Getter, HasManyThroughDefinition} from '../relation.types';
+import {
+  DefaultHasManyThroughRepository,
+  HasManyThroughRepository,
+} from './has-many-through.repository';
+import {Filter, Where} from '../../query';
+
+const debug = debugFactory('loopback:repository:has-many-repository-factory');
+
+export type HasManyThroughRepositoryFactory<
+  Target extends Entity,
+  Through extends Entity,
+  ForeignKeyType
+> = (fkValue: ForeignKeyType) => HasManyThroughRepository<Target, Through>;
+
+export interface AdvancedConstraint<T extends Entity> {
+  dataObject: DataObject<T>;
+  filter: Filter<T>;
+  where: Where<T>;
```

Comment:
Can you add TSDoc comment to explain the purpose of this interface please?

I find it suspicious that we have both a `filter` and a `where` object for the same type here. A `filter` already contains `where` portion, what is the purpose of the other `where` then? (`constraint.filter.where` vs `constraint.where`)

Reply:
I don't technically need to add the filter and where. I will remove the filter since it is confusing. I built the advanced constraint because the code originally consuming the constraint assumed the constraint would be a `DataObject<Entity>`. The hasManyThrough relation needs a constraint built out using `or` in the `where` filter to prevent the software from making too many database queries.

Basically, the following is how the constraint is built given a sourceId.
  1. The junction table is filtered by records with a sourceId
  2. The results are reduced to a where filter using or (this is why I have to make a custom constraint)
  3. The target table is filtered by records that match the where filter

Reply:
This is irrelevant now with the new constraint refactor. Read the following comment.

https://github.com/strongloop/loopback-next/pull/2359#issuecomment-475979639